JoomGalleryfriends/JG4-dev
# Messenger service

## What is this service for
The Messenger service will be used to send messenges to inform users about tasks beeing perfomed within the gallery. This is a direct replacement of the JoomMessenger helper class of JGv3.
This service can send messages eighter via email or via personal messaging system of Joomla. The content of the message is created using the mail templating system of joomla.

## How does it work
When you instantiate the service, you have to select a template and provide the values for the variables to be replaced in the chosen template. Additionally you can select a language in which the message should be sent. For this to work there has to be a mail template created for this specific language. By default only an english mail template is installed.

## How to use this service
You can load the service anywhere in Joomla by doing

```
$component = Factory::getApplication()->bootComponent('com_joomgallery');  // boot JoomGallery component
$component->createMessenger('mail');   // instantiate service by choosing a sending method (available: 'mail','pm')
$msg = $component->getMessenger();   // load the service into your scope
$msg->selectTemplate('com_joomgallery.newimage'); // select the template by providing a template id
$msg->addTemplateData(array(...)); // provide all needed template variables as key value pairs
$msg->send('test@example.com'); // send message to the provided mail adress
```
As you can see, you first need to add the corresponding template before you can send messages with its content. With this service only one template is delivered: 'com_joomgallery.newimage'. This is a template to inform users about a new image beeing added (uploaded) to the gallery.

## How to test the service
**This service need the database to be modified in order to add the new mail template!**

Make sure "Debug System" is "Yes" in Joomla Global configuration.
Open the file `/administrator/com_joomgallery/src/View/Faulties/HtmlView.php` and insert the following code on line 62 after `throw new \Exception(implode("\n", $errors)); }`

### Send a message via email
```
$this->component->createMessenger('mail');
$msg = $this->component->getMessenger();
$msg->selectTemplate('com_joomgallery.newimage');
$msg->addTemplateData(array('user' => '<USERNAME>', 'title' => '<IMAGENAME>', 'category' => '<CATEGORYNAME>'));
$msg->send('<test@example.ch>');
```

### Send a message via personal messaging system of joomla
```
$this->component->createMessenger('pm');
$msg = $this->component->getMessenger();
$msg->selectTemplate('com_joomgallery.newimage');
$msg->addTemplateData(array('user' => '<USERNAME>', 'title' => '<IMAGENAME>', 'category' => '<CATEGORYNAME>'));
$msg->send('<USERID>');
```

### Change content of mail template
Navigate in the Joomla backend to "System/Templates/Mail Templates/JoomGallery: New Image" and change the content of the template. When sending a new message the nes content should be sent...
